### PR TITLE
RUBY 2713 wex charging bo 3 1 5 reverse a payment page

### PR DESCRIPTION
**Pull request by jjromeo**
- [RUBY-2713] Add functionality for reversing payments in waste exemptions engine
- [RUBY-2713] Update waste_exemptions_engine and enhance record reversal functionality

[RUBY-2713]: https://eaflood.atlassian.net/browse/RUBY-2713

### CODE REVIEW

File: app/controllers/record_reversals_controller.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  def index
+    find_resource(params[:registration_reference])
+    @payments = @resource.account.payments.refundable
+  end
+
+  def new
+    setup_form
+    @payment = @resource.account.payments.find_by(id: params[:payment_id])
+    return if @payment
+
```
Comment (PaulDoyle-EA) [inline app/controllers/record_reversals_controller.rb:14]: Should there be an error if the payment is not found?

File: app/presenters/payment_details_presenter.rb [hunk 2]
```diff
@@ -13,28 +13,44 @@ def initialize(registration)
...
+    registration
+      .account
+      .overpaid?
+  end
+
+  def can_display_reversal_link?
+    registration
+      .account
+      .payments
+      .refundable.any?
```
Comment (PaulDoyle-EA) [inline app/presenters/payment_details_presenter.rb]: Are refundable and reversable the same? I would have thought that a partially-reversed payment would be refundable but not reversable.

File: app/services/reverse_payment_service.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    @payment = payment
+    @comments = comments
+    @user = user
+
+    reversal = build_reversal
+    reversal.save!
+
+    true
+  rescue StandardError => e
+    Rails.logger.error "#{e.class} error processing reversal for payment #{payment.id}"
```
Comment (PaulDoyle-EA) [inline app/services/reverse_payment_service.rb]: Also log the error message

File: app/views/payment_details/index.html.erb [hunk 4]
```diff
@@ -165,8 +165,15 @@
...
     <% if can_display_refund_link?(@resource) %>
       <p class="govuk-body">
-        <%= link_to t(".actions_section.links.record_refund"),
-          registration_record_refunds_path(registration_reference: @resource.reference) %>
+      <%= link_to t(".actions_section.links.record_refund"),
+        registration_record_refunds_path(registration_reference: @resource.reference) %>
+      </p>
+    <% end %>
+
+    <% if @presenter.can_display_reversal_link? %>
```
Comment (PaulDoyle-EA) [inline app/views/payment_details/index.html.erb]: This method should be on ActionLinksHelper, not on the presenter

File: app/views/record_refunds/index.html.erb [hunk 5]
```diff
@@ -1,62 +1,11 @@
...
-  </div>
-</div>
+<%= render "shared/payments_table",
+           payments: @payments,
+           resource: @resource,
+           translation_prefix: "record_refunds.index" do |payment| %>
+  <%= link_to new_registration_record_refund_path(reference: @resource.reference, payment_id: payment.id) do %>
+    <%= t("record_refunds.index.refund_payment") %>
+    <span class="govuk-visually-hidden">
+      <%= "ALT" %>
```
Comment (PaulDoyle-EA) [inline app/views/record_refunds/index.html.erb]: What is this for?

File: app/views/record_reversals/index.html.erb [hunk 6]
```diff
@@ -0,0 +1,11 @@
+<%= render "shared/payments_table",
+           payments: @payments,
+           resource: @resource,
+           translation_prefix: "record_reversals.index" do |payment| %>
+  <%= link_to new_registration_record_reversal_path(reference: @resource.reference, payment_id: payment.id) do %>
+    <%= t("record_reversals.index.reverse_payment") %>
+    <span class="govuk-visually-hidden">
+      <%= "ALT" %>
```
Comment (PaulDoyle-EA) [inline app/views/record_reversals/index.html.erb]: And this?

File: app/views/record_reversals/new.html.erb [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        <h2 class="govuk-heading-m"><%= t(".payment_type") %></h2>
+        <p class="govuk-body">
+          <%= t(".payment_types.#{@payment.payment_type}") %>
+        </p>
+      </div>
+
+      <div class="govuk-form-group">
+        <h2 class="govuk-heading-m"><%= t(".amount_to_reverse") %></h2>
+        <p class="govuk-body">
+          £<%= display_pence_as_pounds_and_pence(pence: @payment.payment_amount) %>
```
Comment (PaulDoyle-EA) [inline app/views/record_reversals/new.html.erb]: The "£" should be added by the helper - see `display_pence_as_pounds_sterling_and_pence`.

File: app/views/shared/_payments_table.html.erb [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+          <th class="govuk-table__header">
+            <%= t("#{translation_prefix}.action") %>
+          </th>
+        </tr>
+      </thead>
+      <tbody class="govuk-table__body">
+        <% payments.each do |payment| %>
+          <tr class="govuk-table__row">
+            <td class="govuk-table__cell">
+              <%= payment.created_at.to_date.to_formatted_s(:day_month_year_slashes) %>
```
Comment (PaulDoyle-EA) [inline app/views/shared/_payments_table.html.erb]: Can this go in a presenter? Or is that follow-on work?

File: app/views/shared/_payments_table.html.erb [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+            </td>
+            <td class="govuk-table__cell">
+              <%= payment.reference %>
+            </td>
+            <td class="govuk-table__cell">
+              <%= t("#{translation_prefix}.payment_types.#{payment.payment_type}") %>
+            </td>
+            <td class="govuk-table__cell govuk-table__cell--numeric">
+              £<%= display_pence_as_pounds_and_pence(pence: payment.payment_amount, hide_pence_if_zero: false) %>
```
Comment (PaulDoyle-EA) [inline app/views/shared/_payments_table.html.erb]: As above - see sterling helper

File: db/schema.rb [hunk 10]
```diff
@@ -138,6 +138,15 @@
     t.string "sent_to"
   end
 
+  create_table "companies", force: :cascade do |t|
```
Comment (PaulDoyle-EA) [inline db/schema.rb:151]: What is this for?

File: db/schema.rb [hunk 11]
```diff
@@ -16,7 +16,7 @@
   enable_extension "tsm_system_rows"
 
   create_table "accounts", force: :cascade do |t|
-    t.bigint "registration_id", null: false
```
Comment (PaulDoyle-EA) [inline db/schema.rb:19]: I've seen this locally too, I don't think `null: false` should be dropped.

File: app/views/record_refunds/index.html.erb (hunk 5 above)
Comment (jjromeo) [reply app/views/record_refunds/index.html.erb]: Placeholder for screenreader content, but I guess may not be needed

File: app/views/shared/_payments_table.html.erb (hunk 8 above)
Comment (jjromeo) [reply app/views/shared/_payments_table.html.erb]: Follow-on work

File: app/views/record_reversals/new.html.erb (hunk 7 above)
Comment (jjromeo) [reply app/views/record_reversals/new.html.erb]: Done

File: app/views/record_reversals/index.html.erb (hunk 6 above)
Comment (jjromeo) [reply app/views/record_reversals/index.html.erb]: Removed

File: db/schema.rb [hunk 12]
```diff
@@ -10,13 +10,13 @@
...
 
-ActiveRecord::Schema[7.1].define(version: 2024_11_18_143709) do
+ActiveRecord::Schema[7.1].define(version: 2024_11_28_154543) do
   # These are extensions that must be enabled in order to support this database
   enable_extension "plpgsql"
   enable_extension "tsm_system_rows"
 
   create_table "accounts", force: :cascade do |t|
-    t.bigint "registration_id", null: false
+    t.bigint "registration_id"
```
Comment (PaulDoyle-EA) [inline db/schema.rb:19]: Again, should `null: false` be retained? I keep seeing this locally too.

File: db/schema.rb [hunk 13]
```diff
@@ -148,6 +148,15 @@
     t.string "sent_to"
   end
 
+  create_table "companies", force: :cascade do |t|
+    t.string "name"
```
Comment (PaulDoyle-EA) [inline db/schema.rb:152]: Unclear to me why this change is part of this PR

File: db/schema.rb [hunk 14]
```diff
@@ -157,7 +166,9 @@
     t.text "guidance"
     t.boolean "hidden", default: false
     t.bigint "band_id"
+    t.bigint "waste_activity_id"
```
Comment (PaulDoyle-EA) [inline db/schema.rb:169]: Likewise

File: db/schema.rb [hunk 15]
```diff
@@ -399,6 +413,9 @@
     t.boolean "temp_check_your_answers_flow"
     t.string "temp_company_no"
     t.string "temp_payment_method"
+    t.text "temp_waste_activities", default: [], array: true
```
Comment (PaulDoyle-EA) [inline db/schema.rb:416]: Likewise

File: db/schema.rb [hunk 16]
```diff
@@ -455,6 +472,14 @@
     t.index ["item_type", "item_id"], name: "index_versions_on_item_type_and_item_id"
   end
 
+  create_table "waste_activities", force: :cascade do |t|
```
Comment (PaulDoyle-EA) [inline db/schema.rb:475]: And this

**Review by PaulDoyle-EA — APPROVED** [review]
